pkg/provisioner/controller.go

```diff
@@ -291,6 +293,103 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
+		// updating the OBC reference to the OB and the OBC's phase.
+		ob, err = updateObjectBucketPhase(
+			c.libClientset,
+			ob,
+			v1alpha1.ObjectBucketStatusPhaseBound,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OB %q's status to %q: %v", ob.Name, v1alpha1.ObjectBucketStatusPhaseBound, err)
+		}
+		// update OBC
+		obc.Spec.ObjectBucketName = ob.Name
+		// Do not set obc.Spec.BucketName here: (1) if user has it set, it should be as they set it
+		// (2) if generateBucketName is set, setting this will cause errors if provisioning runs again
+		obc, err = updateClaim(
+			c.libClientset,
+			obc,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OBC: %v", err)
+		}
+		obc, err = updateObjectBucketClaimPhase(
+			c.libClientset,
+			obc,
+			v1alpha1.ObjectBucketClaimStatusPhaseBound,
+			defaultRetryBaseInterval,
+			defaultRetryTimeout)
+		if err != nil {
+			return fmt.Errorf("error updating OBC %q's status to: %v", v1alpha1.ObjectBucketClaimStatusPhaseBound, err)
+		}
+		return nil
+	}
+
+	// recover from degraded state where provisioning is happening again on a completed obc
+	ob, err = getObjectBucketFromClaimKey(key, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+	if err != nil {
+		return fmt.Errorf("error determining if object bucket for this claim exists: %v", err)
+	}
+	if ob != nil {
+		emptyRef := corev1.ObjectReference{}
+		if ob.Spec.ClaimRef == nil || *ob.Spec.ClaimRef == emptyRef || ob.Spec.ClaimRef.UID == types.UID("") {
+			logD.Info("found and existing Object Bucket matching this claim that does not have an object reference, assuming it should belong to this claim, re-associating this claim and stopping provisioning", "object bucket", ob.Name)
+			ob.Spec.ClaimRef, err = claimRefForKey(key, c.libClientset)
+			if err != nil {
+				return fmt.Errorf("error getting reference to OBC: %v", err)
+			}
+			ob, err = updateObjectBucket(ob, c.libClientset, defaultRetryBaseInterval, defaultRetryTimeout)
+			if err != nil {
+				return fmt.Errorf("error updating object bucket with assumed claim ref: %v", err)
+			}
+		} else if obc.UID == ob.Spec.ClaimRef.UID {
+			logD.Info("found an existing Object Bucket belonging to this claim, re-associating this claim and stopping provisioning", "object bucket", ob.Name)
+		} else {
+			return fmt.Errorf("found an object bucket %q that is not owned by this object bucket claim, this is likely an artifact from a previous OBC that did not get cleaned up, not continuing with OBC provisioning", ob.Name)
+		}
+		err = updateOnSuccess()
+		if err != nil {
+			return fmt.Errorf("error re-associating object bucket to claim. %v", err)
+		}
+		// Do not do any more provisioning; the bucket is already provisioned and merely
+		// needed re-associated to the claim
+		logD.Info("re-associated bucket to claim successfully")
```

Comment:
Most of the design doc for this lib refers to what you're calling re-association as _binding_. What do you think about s/re-association/binding/g ?

Reply:
That makes sense. I'll change to `re-bind` instead.